adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,508 @@
+				
+				if (tlsio_instance->sockRead == NULL || tlsio_instance->sockWrite == NULL)
+				{
+					LogError("Unable to create socket streams.")
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+					result = __FAILURE__;
+				}
+				else if (!CFReadStreamSetProperty(tlsio_instance->sockRead, kCFStreamPropertySSLSettings, kCFStreamSocketSecurityLevelTLSv1))
+				{
+					LogError("Unable to start SSL on socket.");
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+					result = __FAILURE__;
+				}
+	            else if (!CFReadStreamOpen(tlsio_instance->sockRead) || !CFWriteStreamOpen(tlsio_instance->sockWrite))
+				{
+					LogError("Failure: connect failure");
+					CFReadStreamClose(tlsio_instance->sockRead);
+					CFWriteStreamClose(tlsio_instance->sockWrite);
+					tlsio_instance->sockRead = NULL;
+					tlsio_instance->sockWrite = NULL;
+					result = __FAILURE__;
+				}
+				else
+				{
+					tlsio_instance->state = TLSIO_APPLEIOS_STATE_OPENING;
+				}
+			}
+		}
+    }
+
+    if (result == 0)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_041: [ If the tlsio_appleios_open get success opening the tls connection, it shall call the tlsio_appleios_dowork. ]*/
+        tlsio_appleios_dowork(tlsio_handle);
+    }
+	
+    return result;
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_043: [ The tlsio_appleios_close shall start the process to close the ssl connection. ]*/
+int tlsio_appleios_close(CONCRETE_IO_HANDLE tlsio_handle, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* on_io_close_complete_context)
+{
+    int result;
+    TLS_IO_INSTANCE* tlsio_instance = (TLS_IO_INSTANCE*)tlsio_handle;
+
+    if (tlsio_handle == NULL)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_049: [ If the tlsio_handle is NULL, the tlsio_appleios_close shall not do anything, and return _LINE_. ]*/
+        LogError("NULL TLS handle.");
+        result = __FAILURE__;
+    }
+    else
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_045: [ The tlsio_appleios_close shall store the provided on_io_close_complete callback function address. ]*/
+        tlsio_instance->on_io_close_complete = on_io_close_complete;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_046: [ The tlsio_appleios_close shall store the provided on_io_close_complete_context handle. ]*/
+        tlsio_instance->on_io_close_complete_context = on_io_close_complete_context;
+
+        if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSED) || (tlsio_instance->state == TLSIO_APPLEIOS_STATE_ERROR))
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_079: [ If the tlsio state is TLSIO_APPLEIOS_STATE_ERROR, or TLSIO_APPLEIOS_STATE_CLOSED, the tlsio_appleios_close shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSE, and return 0. ]*/
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_CLOSED;
+            result = 0;
+        }
+        else if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING) || (tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSING))
```

Comment:
>tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING [](start = 18, length = 53)

On OPENING, please force SSL close, change the state to CLOSED, and return 0 (success).